### DISCUSSION TITLE
Setting action name automatically in Redux Devtools?

### DISCUSSION THREAD

**Discussion by adrenaline681** [pmndrs/zustand#1424] [Q&A]
Currently, if I want to see the actions with proper names inside Redux Devtools i need to define them always making sure i set a proper name like this:

```    
setFilter: (filter)=>set({filter:filter}, false, 'setFilter'),
resetCounter: ()=>set({counter:0}, false, 'resetCounter'),
setFriendsAmount: (friends)=>set({friendsAmount: friends.length}, false, 'setFriendsAmount')
```

It seems to be somewhat unnecessary to keep writing the action name twice. 
Is there a way for the action name in the devtools to be automatically set from the action name? so all i have to do is this:

```    
setFilter: (filter)=>set({filter:filter}),
resetCounter: ()=>set({counter:0}),
setFriendsAmount: (friends)=>set({friendsAmount: friends.length})
```

**Comment by dai-shi** [comment]
That would be tricky with a runtime solution.
A compiler can do, for example, [babel-plugin-macros](https://www.npmjs.com/package/babel-plugin-macros).
Anyone got interested to develop a 3rd-party library?

**Comment by Janaka-Steph** [comment]
Fixing this would be really cool. The only issue would be when having multiple set in one function they would have the same name.

But there is a confusion I don't get. We should log an action when an action is called, not when `set` is called.
I have actions that don't call any `set` but return values instead.

**Reply by ecatugy** [reply]
Has anyone figured out how to set the name of the action in Zustand with Redux Devtools?

**Reply by adrenaline681** [reply]
Not automatically, but you can set it manually like this:

```
  setFilterDistance: (distance)=>{
      set((state)=>{
          state.filters.distance = distance
      }, false, 'setFilterDistance')
  },
```

Im a coding noob but yea, that's how I do it :D

**Reply by ecatugy** [reply]
> Not automatically, but you can set it manually like this:
> 
> ```
>   setFilterDistance: (distance)=>{
>       set((state)=>{
>           state.filters.distance = distance
>       }, false, 'setFilterDistance')
>   },
> ```
> 
> Im a coding noob but yea, that's how I do it :D

I'm using the slices pattern with TypeScript and I was able to do it using immer. Here's an example:

```typescript
import { create, StateCreator } from 'zustand';
import {
  Connection,
  Edge,
  EdgeChange,
  Node,
  NodeChange,
  addEdge,
  OnNodesChange,
  OnEdgesChange,
  OnConnect,
  applyNodeChanges,
  applyEdgeChanges,
} from 'reactflow';
import { StoreState } from '../store';
import initialNodes from './nodes';
import initialEdges from './edges';

import produce from 'immer';

export type RFStateSlice = {
  nodes: Node<NodeData>[];
  edges: Edge[];
  onNodesChange: OnNodesChange;
  onEdgesChange: OnEdgesChange;
  onConnect: OnConnect;
  updateNodeColor: (nodeId: string, color: string) => void;
};

export type NodeData = {
  color: string;
};

export const createFlowSlice: StateCreator<
  StoreState,
  [['zustand/persist', unknown], ['zustand/devtools', never]],
  [],
  RFStateSlice
> = (set, get, api) => ({
  nodes: initialNodes,
  edges: initialEdges,
  onNodesChange: (changes: NodeChange[]) => {
    set(
      produce((state) => {
        state.nodes = applyNodeChanges(changes, state.nodes);
      }),
      false,
      'onNodesChange'
    );
  },
  onEdgesChange: (changes: EdgeChange[]) => {
    set(
      produce((state) => {
        state.edges = applyEdgeChanges(changes, state.edges);
      }),
      false,
      'onEdgesChange'
    );
  },
  onConnect: (connection: Connection) => {
    set(
      produce((state) => {
        state.edges = addEdge(connection, state.edges);
      }),
      false,
      'onConnect'
    );
  },
  updateNodeColor: (nodeId: string, color: string) => {
    set(
      produce((state) => {
        state.nodes = state.nodes.map((node: any) => {
          if (node.id === nodeId) {
            node.data.color = color;
          }
          return node;
        });
      }),
      false,
      'updateNodeColor'
    );
  },
});
